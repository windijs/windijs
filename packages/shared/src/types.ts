// TODO: we need typescript 4.8 to support infer number from template
export type ParseInt<T> = T extends `${infer N extends number}` ? N : never

export type Negative<T extends object> = {
  [k in keyof T as k extends 0 | "0" ? k : k extends number ? ParseInt<`-${k}`> : `-${string & k}`]: T[k] extends number | string ? `-${T[k]}` : T[k]
};

export type NonNegativeInteger<T extends number> =
  number extends T
  ? never
  : `${T}` extends `-${string}` | `${string}.${string}`
  ? never
  : T;

// Range Type
// https://www.codegrepper.com/code-examples/typescript/typescript+type+number+in+range

type PrependNextNum<A extends Array<unknown>> = A["length"] extends infer T ? ((t: T, ...a: A) => void) extends ((...x: infer X) => void) ? X : never : never;

type EnumerateInternal<A extends Array<unknown>, N extends number> = { 0: A, 1: EnumerateInternal<PrependNextNum<A>, N> }[N extends A["length"] ? 0 : 1];

export type Enumerate<N extends number> = EnumerateInternal<[], N> extends (infer E)[] ? E : never;

export type Range<FROM extends number, TO extends number> = Exclude<Enumerate<TO>, Enumerate<FROM>>

export type Fraction<T extends number> = `${Range<1, T> & number}/${T}`

export type FractionObject<FROM extends number, TO extends number> = {
  [key in Range<FROM, TO> as `${Fraction<key & number>}`]: string
}

export type RangeTuple<FROM extends number, TO extends number> = TuplifyUnion<Range<FROM, TO>>

// Convert Union Type to Tuple Type
// https://stackoverflow.com/questions/55127004/how-to-transform-union-type-to-tuple-type/55128956#55128956

type UnionToIntersection<U> =
  (U extends any ? (k: U) => void : never) extends ((k: infer I) => void) ? I : never
type LastOf<T> =
  UnionToIntersection<T extends any ? () => T : never> extends () => (infer R) ? R : never

type Push<T extends any[], V> = [...T, V];

export type TuplifyUnion<T, L = LastOf<T>, N = [T] extends [never] ? true : false> =
  true extends N ? [] : Push<TuplifyUnion<Exclude<T, L>>, L>
