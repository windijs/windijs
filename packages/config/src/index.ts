export * from "./bg";
export * from "./svg";
export * from "./flex";
export * from "./grid";
export * from "./border";
export * from "./effects";
export * from "./behaviors";
export * from "./filter";
export * from "./opacity";
export * from "./sizing";
export * from "./spacing";
export * from "./animation";
export * from "./transform";
export * from "./positioning";
export * from "./interactivity";
export * from "./transitions";
export * from "./typography";
export * from "./keyframes";
export * from "./interface";
