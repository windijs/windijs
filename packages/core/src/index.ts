export * from "./base";
export * from "./api";
export * from "./handler";
export * from "./prefixer";
export * from "./builder";
