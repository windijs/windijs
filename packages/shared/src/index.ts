export * from "./gen";
export * from "./utils";
export * from "./types";
