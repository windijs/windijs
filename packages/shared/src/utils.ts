import type { FractionObject, Negative, RangeTuple } from "./types";

/**
 * Check if value is a number
 * @param value {string}
 * @returns {boolean}
 */
export function isNumber (value: string) {
  // @ts-ignore
  return !isNaN(value);
}

export function isFraction (amount: string): boolean {
  return /^\d+\/\d+$/.test(amount);
}

export function isSize (amount: string): boolean {
  return /^-?(\d+(\.\d+)?)+(rem|em|px|rpx|vh|vw|ch|ex|cm|mm|in|pt|pc)$/.test(amount);
}

export function roundUp (num: number, precision = 0): number {
  precision = Math.pow(10, precision);
  return Math.round(num * precision) / precision;
}

export function fracToPercent (amount: string): string | undefined {
  const matches = amount.match(/[^/]+/g);
  if (!matches || matches.length < 2) return;
  const a = +matches[0];
  const b = +matches[1];
  return roundUp((a / b) * 100, 6) + "%";
}

export const hasKey = <T extends object>(obj: T, k: keyof any): k is keyof T => k in obj;

export function commaJoin (...items: (string|number|undefined)[]) {
  return items.filter(i => i).join(", ");
}

export function parenWrap (key: string, value: string) {
  return key + "(" + value + ")";
}

export function camelToDash (str: string) {
  return str.replace(/([A-Z])/g, v => `-${v.toLowerCase()}`);
}

export function dashToCamel (str: string) {
  return str.replace(/(-[a-z])/g, v => v.toUpperCase().replace("-", ""));
}

export function hash (str: string): string {
  str = str.replace(/\r/g, "");
  let hash = 5381;
  let i = str.length;

  while (i--) hash = ((hash << 5) - hash) ^ str.charCodeAt(i);
  return (hash >>> 0).toString(36);
}

export function indent (value: string, count: number = 2) {
  return " ".repeat(count) + value;
}

export const entries = <T extends object | Map<string, any>> (t: T) => t instanceof Map ? t.entries() : Object.entries(t);

export function negative <T extends Record<string | number, string>> (t: T): Negative<T> {
  return Object.fromEntries(Object.entries(t).filter(([, v]) => !/^0(px|rem|em)?$/.test(v)).map(([k, v]) => ["-" + k, "-" + v])) as Negative<T>;
}

export function range<S extends number, E extends number> (start: S, end: E): RangeTuple<S, E> {
  return new Array(end - start).fill(0).map((_, i) => i + start) as RangeTuple<S, E>;
}

export function fractions<S extends number, E extends number> (start: S, end: E) {
  return Object.fromEntries((range(start, end) as number[]).map(i => range(1, i as 0).map(v => [v + "/" + i, v / i * 100] as [string, number]).map(([k, v]) => [k, v % 1 === 0 ? v + "%" : v.toFixed(6) + "%"])).flat()) as FractionObject<S, E>;
}

export function spacings<T extends Array<any>> (numbers: T) {
  return Object.fromEntries(numbers.map(i => [i, i / 4 + (i === 0 ? "px" : "rem")])) as Record<T[number], string>;
}

export function degrees<T extends number> (numbers: T[]) {
  return Object.fromEntries(numbers.map(i => [i, i + "deg"])) as Record<T, string>;
};

export function scales<T extends number> (numbers: T[]) {
  return Object.fromEntries(numbers.map(i => [i, i / 100 + ""])) as Record<T, string>;
};

export function omit<T extends object, K extends object> (t: T, keys: K) {
  return Object.fromEntries(Object.entries(t).filter(([k]) => !(k in keys))) as Omit<T, keyof K>;
}
