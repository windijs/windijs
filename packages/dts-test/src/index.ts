import { defineConfig } from "@windijs/config";

defineConfig({
  utilities: {

  },
});
