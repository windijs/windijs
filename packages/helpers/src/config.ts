import { Handler, StyleLoader, StyleObject } from "./types";

import { Theme } from "@windijs/config";

export type DarkModeConfig = "class" | "media" | false;

export type UtilitiesConfig = { [key: string]: Handler<unknown> | Array<Handler<unknown>> | StyleObject | UtilitiesConfig };

export type VariantsConfig = { [key: string]: string | string[] };

export interface Config<T extends object = Theme> {
  // presets?: Config[];
  // prefixer?: boolean;
  // attributify?: boolean | {
  //   prefix?: string
  //   separator?: string
  //   disable?: string[]
  // }
  separator?: string;
  important?: boolean | string;
  darkMode?: DarkModeConfig;
  theme?: T;
  // variantOrder?: string[];
  // plugins?: Plugin[];
  // handlers?: Handlers;
  // corePlugins?: (keyof BaseTheme)[] | string[] | { [ T in keyof BaseTheme ] : boolean } | { [ key:string ] : boolean };
  // prefix?: string;
  // exclude?: RegExp[];
  // alias?: { [key:string]: string };
  // shortcuts?: { [key:string]: Shortcut };
  // purge?: unknown;
  utilities?: UtilitiesConfig;
  variants?: VariantsConfig;
  styleLoader?: StyleLoader;
  [key:string]: any;
}

export function defineConfig (config: Config) {
  return config;
}
