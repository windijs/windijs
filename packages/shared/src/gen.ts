import { indent, isNumber, isVarName } from "./utils";

/**
 * Generate type interface for config object
 */
export function dtsConfig<T extends object> (obj: T, defaultType = "string", ignores: string[] = [], indentLevel = 0, indentation = 4): string {
  const contents = Object.entries(obj).filter(([k]) => !ignores.includes(k)).map(([k, v]) => (isVarName(k) || isNumber(k) ? k : JSON.stringify(k)) + ": " + (typeof v === "object" && v != null && !Array.isArray(v) ? dtsConfig(v, defaultType, ignores, indentLevel + 1, indentation) : defaultType + ";"));
  return "{\n" + contents.map(i => indent(i, indentation * (indentLevel + 1))).join("\n") + "\n" + indent(indentLevel === 0 ? "}" : "};", indentation * indentLevel);
}

/**
 * Generate utilities.d.ts with config
 */
export function genUtilitiesDts<T extends object> (tmpl: string, config: T): string {
  let code = Object.entries(config).reduce((prev, [k, v]) => prev.replace(new RegExp(`"\\$windi\\.config\\.${k}Config.proxy"`, "g"), dtsConfig(v, "StyleObject<{}>")), tmpl);
  if ("colors" in config) {
    code = code
      // @ts-ignore
      .replace(/"\$windi\.config\.colorsConfig"/g, dtsConfig(config.colors))
      // @ts-ignore
      .replace(/"\$windi\.color\.colors\.proxy"/g, dtsConfig(config.colors, "StyleObject<{opacity: (op: number) => StyleObject<{readonly gradient: StyleObject<{}>}>; readonly gradient: StyleObject<{}>}>"));
  }
  return code;
}

/**
 * Generate utilities.js with config
 */
export function genUtilitiesJs<T extends object> (tmpl: string, config: T): string {
  const start = tmpl.match(/const\s+\S+\s*=/)?.index ?? 0;

  let code = Object.keys(config).reduce((prev, k) => prev.replace(new RegExp(`(?<=configHandler\\().*?${k}Config`, "g"), k + "WindiConfigInject"), tmpl.slice(start));
  // @ts-ignore
  if ("colors" in config) code = code.replace(/const\s+colors\s*=[\s\S]+?;/, "const colors = colorsWindiConfigInject;");

  return [tmpl.slice(0, start), ...Object.entries(config).map(([k, v]) => `const ${k}WindiConfigInject = ${JSON.stringify(v)};`), code].join("\n");
}

/**
 * Generate utilities.mjs with config
 */
export const genUtilitiesMjs = genUtilitiesJs;
