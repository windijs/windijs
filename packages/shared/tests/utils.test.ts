import { backgroundSizeConfig, borderRadiusConfig, touchActionConfig } from "@windi/config";

import { dtsConfig } from "../src";

test("dts", () => {
  expect(dtsConfig(backgroundSizeConfig)).toMatchSnapshot();
});

test("dts nested", () => {
  expect(dtsConfig(touchActionConfig)).toMatchSnapshot();
  expect(dtsConfig({
    a: {
      b: {
        c: {
          d: "d",
        },
      },
    },
  })).toMatchSnapshot();
});

test("dts number and string", () => {
  expect(dtsConfig({
    $: "a",
    2: "b",
    1.5: "c",
    "dts-test": "d",
  })).toMatchSnapshot();
});

test("dts with default type", () => {
  expect(dtsConfig(touchActionConfig, "StyleObject<{}>")).toMatchSnapshot();
});

test("dts ignore DEFAULT", () => {
  expect(dtsConfig(borderRadiusConfig, "StyleObject<{}>", ["DEFAULT"])).toMatchSnapshot();
});

test("dst with array should be replaced", () => {
  expect(dtsConfig({ a: ["a", "b"] })).toMatchSnapshot();
});
