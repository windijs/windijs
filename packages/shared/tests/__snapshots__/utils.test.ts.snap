// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`dst with array should be replaced 1`] = `
"{
    a: string;
};"
`;

exports[`dts 1`] = `
"{
    auto: string;
    cover: string;
    contain: string;
};"
`;

exports[`dts ignore DEFAULT 1`] = `
"{
    1: StyleObject<{}>;
    none: StyleObject<{}>;
    sm: StyleObject<{}>;
    md: StyleObject<{}>;
    lg: StyleObject<{}>;
    xl: StyleObject<{}>;
    xxl: StyleObject<{}>;
    xxxl: StyleObject<{}>;
    full: StyleObject<{}>;
};"
`;

exports[`dts nested 1`] = `
"{
    auto: string;
    none: string;
    pan: {
        x: string;
        y: string;
        left: string;
        right: string;
        up: string;
        down: string;
    };
    \\"pinch-zoom\\": string;
    manipulation: string;
};"
`;

exports[`dts nested 2`] = `
"{
    a: {
        b: {
            c: {
                d: string;
            };
        };
    };
};"
`;

exports[`dts number and string 1`] = `
"{
    2: string;
    $: string;
    1.5: string;
    \\"dts-test\\": string;
};"
`;

exports[`dts with default type 1`] = `
"{
    auto: StyleObject<{}>;
    none: StyleObject<{}>;
    pan: {
        x: StyleObject<{}>;
        y: StyleObject<{}>;
        left: StyleObject<{}>;
        right: StyleObject<{}>;
        up: StyleObject<{}>;
        down: StyleObject<{}>;
    };
    \\"pinch-zoom\\": StyleObject<{}>;
    manipulation: StyleObject<{}>;
};"
`;
